File: appunvs/ios/Runtime/Net/BoxEventsClient.swift
// BoxEventsClient — long-lived SSE consumer for GET /box/events.
//
// Opened once at sign-in (RuntimeApp.SignedInRoot.wireAndLoad) and torn
// down at sign-out.  Receives `bundle_ready` events for any box owned by
// the authenticated user; the consumer typically calls
// `boxStore.refresh()` so Stage's reactive binding to activeBox.bundleURL
// flips and RuntimeView re-mounts the new bundle.
//
// Reconnect: this client keeps trying.  Network drops, server restarts,
// proxy timeouts — all are handled by an internal exponential-backoff
// reconnect loop.  On a fresh (re)connect we yield a `.reconnected`
// element so the caller can `boxStore.refresh()` and pick up any events
// missed while disconnected (publishes during the gap fan out to
// subscribers that don't exist; we recover by polling once).
//
// Heartbeats (`event: heartbeat`) keep the TCP connection alive through
// reverse proxies; they're consumed silently and never reach the caller.
//
// Surface: `events()` returns an AsyncStream the caller drives via a
// `for await` loop inside a `.task { }` modifier.  Cancellation
// propagates: when the consuming Task is cancelled (View disappears,
// .task replaces), AsyncStream's onTermination fires and the internal
// reconnect Task cancels.  No explicit stop() needed — Task lifecycle
// owns it.
import Foundation

struct BoxBundleReadyEvent: Decodable, Sendable {
    let type: String
    let boxID: String
    let version: String
    let uri: String
    let contentHash: String
    let sizeBytes: Int64

    enum CodingKeys: String, CodingKey {
        case type
        case boxID       = "box_id"
        case version
        case uri
        case contentHash = "content_hash"
        case sizeBytes   = "size_bytes"
    }
}

/// What the events() stream yields.
enum BoxStreamEvent: Sendable {
    case bundleReady(BoxBundleReadyEvent)
    /// Emitted on every successful (re)connect AFTER the first.  The
    /// caller uses this as a hint to re-fetch the box list and catch up
    /// on events that may have fired while disconnected.
    case reconnected
}

/// Plain final class — same rationale as AISSEClient: no mutable state
/// the loop needs synchronized access to, and going through an actor
/// would force a hop on every byte read.
final class BoxEventsClient: @unchecked Sendable {
    private let baseURL: URL
    private let session: URLSession
    private let tokenProvider: @Sendable () -> String?

    /// Backoff schedule for reconnect attempts.  Walks the array and
    /// stays at the last value indefinitely.  6 entries totals ~62s
    /// before settling at the 30s cap.
    private let backoff: [TimeInterval] = [1, 2, 5, 10, 20, 30]

    init(
        baseURL: URL = NetConfig.relayBaseURL,
        session: URLSession = .shared,
        tokenProvider: @escaping @Sendable () -> String?
    ) {
        self.baseURL = baseURL
        self.session = session
        self.tokenProvider = tokenProvider
    }

    /// Returns an infinite stream of bundle_ready / reconnected events.
    /// The internal reconnect Task lives until the consumer cancels its
    /// for-await loop (e.g. by Task cancellation on view disappear).
    func events() -> AsyncStream<BoxStreamEvent> {
        AsyncStream { continuation in
            let task = Task { [self] in
                var attempt = 0
                var firstConnect = true
                while !Task.isCancelled {
                    let connected = await runOnce(
                        firstConnect: firstConnect,
                        continuation: continuation,
                    )
                    if Task.isCancelled { break }
                    if connected {
                        attempt = 0
                        firstConnect = false
                    } else {
                        attempt = min(attempt + 1, backoff.count - 1)
                    }
                    let delay = backoff[min(attempt, backoff.count - 1)]
                    try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// One open-stream-and-read attempt.  Returns true if it managed to
    /// connect (so the caller can reset its backoff counter), false if
    /// the request failed before any data arrived.
    private func runOnce(
        firstConnect: Bool,
        continuation: AsyncStream<BoxStreamEvent>.Continuation
    ) async -> Bool {
        var req = URLRequest(url: baseURL.appendingPathComponent("/box/events"))
        req.httpMethod = "GET"
        req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        if let token = tokenProvider() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        do {
            let (bytes, response) = try await session.bytes(for: req)
            guard let http = response as? HTTPURLResponse,
                  (200..<300).contains(http.statusCode) else {
                return false
            }
            // Tell the consumer we're (re)connected — but only after
            // the first connect, so the very first refresh isn't
            // duplicated against the wireAndLoad initial refresh.
            if !firstConnect {
                continuation.yield(.reconnected)
            }

            var event: String = "message"
            for try await line in bytes.lines {
                if Task.isCancelled { break }
                if line.isEmpty {
                    event = "message"
                    continue
                }
                if line.hasPrefix("event: ") {
                    event = String(line.dropFirst(7))
                    continue
                }
                if line.hasPrefix("data: ") {
                    let json = String(line.dropFirst(6))
                    if let value = parse(event: event, json: json) {
                        continuation.yield(value)
                    }
                }
            }
            return true
        } catch {
            return false
        }
    }

    private func parse(event: String, json: String) -> BoxStreamEvent? {
        switch event {
        case "bundle_ready":
            guard let data = json.data(using: .utf8) else { return nil }
            if let decoded = try? JSONDecoder().decode(BoxBundleReadyEvent.self, from: data) {
                return .bundleReady(decoded)
            }
            return nil
        case "heartbeat", "message":
            return nil
        default:
            return nil
        }
    }
}

File: appunvs/android/app/src/main/java/com/appunvs/runtime/net/BoxEventsClient.kt
// BoxEventsClient — long-lived SSE consumer for GET /box/events.
//
// Opened once at sign-in (MainActivity LaunchedEffect) and torn down at
// sign-out / process exit.  Receives `bundle_ready` events for any box
// owned by the authenticated user; the consumer typically calls
// `boxRepo.refresh()` so Stage's reactive binding to activeBox.bundleURL
// flips and RuntimeView re-mounts the new bundle.
//
// Reconnect: this client keeps trying.  Network drops, server restarts,
// proxy timeouts — all are handled by an internal exponential-backoff
// reconnect loop.  On a fresh (re)connect we emit a `Reconnected` event
// so the consumer can `boxRepo.refresh()` and pick up any events missed
// while disconnected (publishes during the gap fan out to subscribers
// that don't exist; we recover by polling once).
//
// Heartbeats (`event: heartbeat`) keep the TCP connection alive through
// reverse proxies; they're consumed silently and never reach the caller.
//
// Surface: `events()` returns a cold `Flow` the caller drives via
// `collect { }`.  Cancellation via the collecting coroutine's scope
// cancels the in-flight HTTP call (OkHttp's response.use propagates).
package com.appunvs.runtime.net

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.coroutines.coroutineContext
import okhttp3.OkHttpClient
import okhttp3.Request

@Serializable
data class BoxBundleReadyEvent(
    val type: String,
    val box_id: String,
    val version: String,
    val uri: String,
    val content_hash: String,
    val size_bytes: Long,
)

/// What the events() Flow emits.
sealed class BoxStreamEvent {
    data class BundleReady(val payload: BoxBundleReadyEvent) : BoxStreamEvent()
    /// Emitted on every successful (re)connect AFTER the first.  The
    /// caller uses this as a hint to re-fetch the box list and catch up
    /// on events that may have fired while disconnected.
    object Reconnected : BoxStreamEvent()
}

class BoxEventsClient(
    private val http: OkHttpClient,
    private val baseURL: String = NetConfig.relayBaseURL,
    private val json: Json = RelayClient.json,
) {
    /// Backoff schedule for reconnect attempts, in milliseconds.  Walks
    /// the array and stays at the last value indefinitely.  6 entries
    /// totals ~62s before settling at the 30s cap.
    private val backoffMs = longArrayOf(1_000, 2_000, 5_000, 10_000, 20_000, 30_000)

    fun events(): Flow<BoxStreamEvent> = flow {
        var attempt = 0
        var firstConnect = true
        while (true) {
            val connected = runOnce(firstConnect, this)
            if (connected) {
                attempt = 0
                firstConnect = false
            } else {
                attempt = (attempt + 1).coerceAtMost(backoffMs.size - 1)
            }
            delay(backoffMs[attempt.coerceAtMost(backoffMs.lastIndex)])
        }
    }.flowOn(Dispatchers.IO)

    /// One open-stream-and-read attempt.  Returns true if the response
    /// hit an OK status so the caller can reset its backoff counter.
    /// Errors during read mid-stream still count as connected (the
    /// stream just dropped — backoff resets next iteration).
    ///
    /// Cancellation: OkHttp's execute()/readUtf8Line() are blocking
    /// and don't observe coroutine cancellation on their own.  We bind
    /// call.cancel() to the producing coroutine's Job — when the
    /// LaunchedEffect that hosts our flow cancels, the job completes,
    /// the handler fires, and the in-flight HTTP request aborts
    /// (closing the socket; readUtf8Line throws IOException; the
    /// catch below funnels back through the flow with no crash).
    private suspend fun runOnce(
        firstConnect: Boolean,
        emitter: kotlinx.coroutines.flow.FlowCollector<BoxStreamEvent>,
    ): Boolean {
        val req = Request.Builder()
            .url("$baseURL/box/events")
            .header("Accept", "text/event-stream")
            .get()
            .build()

        val call = http.newCall(req)
        val handle = coroutineContext[Job]?.invokeOnCompletion { call.cancel() }
        return try {
            call.execute().use { resp ->
                if (!resp.isSuccessful) return false
                val source = resp.body?.source() ?: return false

                if (!firstConnect) {
                    emitter.emit(BoxStreamEvent.Reconnected)
                }

                var event = "message"
                while (!source.exhausted()) {
                    val line = source.readUtf8Line() ?: break
                    if (line.isEmpty()) {
                        event = "message"
                        continue
                    }
                    if (line.startsWith("event: ")) {
                        event = line.removePrefix("event: ")
                        continue
                    }
                    if (line.startsWith("data: ")) {
                        val payload = line.removePrefix("data: ")
                        parse(event, payload)?.let { emitter.emit(it) }
                    }
                }
                true
            }
        } catch (ce: CancellationException) {
            // Coroutine cancellation must propagate — don't swallow.
            throw ce
        } catch (_: Throwable) {
            false
        } finally {
            handle?.dispose()
        }
    }

    private fun parse(event: String, payload: String): BoxStreamEvent? {
        return when (event) {
            "bundle_ready" -> runCatching {
                BoxStreamEvent.BundleReady(json.decodeFromString(BoxBundleReadyEvent.serializer(), payload))
            }.getOrNull()
            "heartbeat", "message" -> null
            else -> null
        }
    }
}
